src/colmap/estimators/alignment.cc

```diff
@@ -456,15 +455,28 @@ bool MergeReconstructions(const double max_reproj_error,
 
   // Register the missing images in this src_reconstruction.
   for (const auto image_id : missing_image_ids) {
-    auto src_image = src_reconstruction.Image(image_id);
-    src_image.ResetCameraPtr();
-    src_image.SetCamFromWorld(
-        TransformCameraWorld(tgt_from_src, src_image.CamFromWorld()));
-    if (!tgt_reconstruction.ExistsCamera(src_image.CameraId())) {
+    const auto& src_image = src_reconstruction.Image(image_id);
+    auto tgt_image = src_image;
+    tgt_image.ResetCameraPtr();
+    tgt_image.ResetFramePtr();
+    if (!tgt_reconstruction.ExistsRig(src_image.FramePtr()->RigId())) {
+      tgt_reconstruction.AddRig(
+          src_reconstruction.Rig(src_image.FramePtr()->RigId()));
+    }
+    if (!tgt_reconstruction.ExistsCamera(tgt_image.CameraId())) {
       tgt_reconstruction.AddCamera(
-          src_reconstruction.Camera(src_image.CameraId()));
+          src_reconstruction.Camera(tgt_image.CameraId()));
+    }
+    if (!tgt_reconstruction.ExistsFrame(tgt_image.FrameId())) {
+      tgt_reconstruction.AddFrame(
+          src_reconstruction.Frame(tgt_image.FrameId()));
     }
-    tgt_reconstruction.AddImage(src_image);
+    tgt_image.SetFramePtr(&tgt_reconstruction.Frame(tgt_image.FrameId()));
+    const Rigid3d cam_from_tgt_world =
+        TransformCameraWorld(tgt_from_src, src_image.CamFromWorld());
+    tgt_reconstruction.Frame(tgt_image.FrameId())
+        .SetCamFromWorld(tgt_image.CameraId(), cam_from_tgt_world);
+    tgt_reconstruction.AddImage(tgt_image);
     tgt_reconstruction.RegisterImage(image_id);
```

Comment:
Consider splitting this code block into functions `MergeReconstructions(src_reconstruction, tgt_reconstruction, tgt_from_src)` and `CopyImageFromReconstruction(src_reconstruction, tgt_reconstruction, tgt_from_src, image_id)` that can be more easily tested later on - can be done in a later PR to keep the diffs clean.